Added >=0 check on delta

nuvious opened the pull request:
## References

Alleviates #16254

## Code changes

Added check on delta to see if it is positive. If it is, return 0 for value and 'seconds' for the unit per documentation below and convention demonstrated at line [48](https://github.com/jupyterlab/jupyterlab/blob/main/packages/coreutils/src/time.ts#L48)/52  (current/proposed time.ts implementation).

Without this check if delta is greater than 0, `Math.ceil(delta / unit.milliseconds)` will compute to 1 with the units being "years" in the first check resulting in the behavior outlined in #16254

## User-facing changes

User will se 'now' or '0 seconds'. Had trouble building/installing jupterlab from source in a fresh venv and still trying to so I can provide manual testing validation

## Backwards-incompatible changes

N/A

nuvious committed:
Added >=0 check on delta

Added check on delta to see if it is positive. If it is, return 0 for value and 'seconds' for the unit per documentation below and convention demonstrated at line 48/52  (current/proposed time.ts implementation).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat/format

JasonWeill commented:
@nuvious Thank you so much for opening this issue and pull request! It is possible, using commands like `touch`, to manually set a date in the future for a particular file. It looks like your case covers a situation where the delta is only slightly larger than 0 (by a second or so) and is being misreported as 1 year from now. There are other ways to handle this:

1. Treat files with a modified date within a few seconds of now as "now" (common on many hosted software products)
2. Revising the logic around the `ceil` function to handle cases where the file's datestamp is in the future.

I don't think that treating _all_ future-dated files as being modified "now" is the best approach to address #16254.

It would also be good to improve the unit test coverage for future-dated files.

nuvious commented:
I think it is reasonable to just treat anything resolving to the future as "now". This is a narrow edge case and the value is "Last Modified" which implies a context in the past. Last modified should never be in the future because that doesn't make sense in any context. The >=0 check acts as an input validation for the ceiling function for this edge case. That's all that's really needed here is some input validation.

I can try my hands at the unit tests but need to familiarize myself with the backend a bit. Will look into that when I get a chance. Also have started using the docker/start.sh environment builder as well to do manual testing and should be able to provide that info in the near future.

krassowski commented:
Any updates @nuvious?

nuvious commented:
Hey @krassowski, looking at it again today between work and school stuff. Got the development environment up today and did some manual tests to evaluate if the fix is working. Going to try to get into the test suite stuff today; not sure why my changes would result in the above failures, but I can dig a little deeper as I familiarize myself with the test pipelines.

nuvious committed:
Update galata.yml

nuvious committed:
Update galata.yml

Didn't need to do this, can just merge to main in my own branch to kick of testing CICD

JasonWeill commented on packages/coreutils/src/time.ts:
```diff
@@ -36,6 +36,10 @@ export namespace Time {
     const delta = new Date(value).getTime() - Date.now();
+    // Delta may be positive due to async execution; default to 0 seconds if it is
+    if (delta >= 0) {
+      return formatter.format(0, "seconds");
```
```suggestion
      return formatter.format(0, 'seconds');
```

This would fix the lint error reported in the pipeline.

afshin committed:
Update packages/coreutils/src/time.ts

Co-authored-by: Jason Weill <93281816+JasonWeill@users.noreply.github.com>

afshin committed:
Merge branch 'main' into 16254-FixLastModifiedOnCreateSave

nuvious committed:
Merge branch 'main' into 16254-FixLastModifiedOnCreateSave

nuvious commented:
@krassowski, dusted this off after a comment got left on the referenced issue. Got the devcontainer up and running and tested my changes after a merge with the upstream main. Going to test it again locally some more but just letting the CICD run to see what fails. I noticed a screenshot failed the last run and I'm going to look into that one specifically if it fails. May need some guidance on updating the screenshot for the test if the referenced issue it the reason the test failed.

nuvious commented:
P.S. I had to add `git config --global --add safe.directory /workspaces/jupyterlab` to the postCreateCommand in the devcontainer definition before the micromamba call as well as as the first command run in the final `RUN` command in the Dockerfile to get the dev container to work. May be missing something about local dev using the devcontainer, but the build process was erroring out on untrusted directories on some git command calls.

nuvious commented:
@JasonWeill / @krassowski, have a few alternative proposals if the simple >=0 check is still not desired.

First is to consider anything within 5 seconds to be defaulted to 0 seconds or 'now':

```typescript
    // Delta may be positive due to async execution; consider anything less than 5 seconds to be 'now'
    if (Math.abs(delta) < 5) {
      return formatter.format(0, 'seconds');
    }
```

Second is to adjust the calculation of amount to be execute a `Math.min` on the amount calculation so if it's positive it's set to 0 as an alternative to the if-statement based solution already in the queue.

```typescript
    const amount = Math.floor(Math.ceil(delta / unit.milliseconds), 0);
```

Would prefer the later solution over the former since any delta that projects into the future is likely an error. It's possible someone may have their system clock set wrong, but in that context the delta since last save is completely unreliable at that point. Think it's elegant to just treat any positive delta as 'now' in the context of displaying how long since the last edit to the user.

nuvious commented:
Just following up on this I'm still unsure of how to test this. Is there a chance we could close this out and then open up a different "find a better solution to" issue? I'm unfamiliar with how I could force this discrepancy in testing. Still standing by for any guidance on the above questions when you have time.
